File: src/lua/fs.lua
local _ = {}

---@param filename string
---@return string | nil
function _:read(filename)
    local file = io.open(filename, "r+")
    if not file then
        error("no file")
    end

    local data = file:read("a")
    file:close();
    return data
end

---@param filename string
---@param data string
function _:write(filename, data)
    local file = io.open(filename, "w")
    if not file then
        error("cannot open file")
    end

    local wrote = file:write(data) ~= nil
    file:close()
    if not wrote then
        error("failed to write to file")
    end
end

return _

File: src/fs.ts
import { Result } from "./result";

export type WriteMode = "w" | "a";

export function writeFile(
  path: string,
  mode: WriteMode,
  data: string,
): Result<void, string> {
  const [file, errStr] = io.open(path, mode);

  if (!file) {
    return {
      ok: false,
      error: errStr,
    };
  }

  const [written, writtenErrStr] = file.write(data);
  if (!written) {
    return {
      ok: false,
      error: writtenErrStr,
    };
  }

  file.close();

  return {
    ok: true,
    value: undefined,
  };
}

export function readFile(path: string): Result<string, string> {
  const [file, errStr] = io.open(path, "r");

  if (!file) {
    return {
      ok: false,
      error: errStr,
    };
  }

  const data = file.read("a");
  if (!data) {
    return {
      ok: false,
      error: "No data read",
    };
  }

  return {
    ok: true,
    value: data,
  };
}
